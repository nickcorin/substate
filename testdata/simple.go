package testdata

//go:generate gensubstate

type Substate interface {
	FooClient() FooClient
	BarClient() BarClient
}

type FooClient interface{}

type BarClient interface{}
