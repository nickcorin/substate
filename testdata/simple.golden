// Code generated by gensubstate at testdata/simple.go; DO NOT EDIT.

package testdata

import (
	"testing"
)

func NewSubstate(injectors ...Injector) *substate {
	var s substate

	for _, injector := range injectors {
		injector.Inject(&s)
	}

	return &s
}

type Injector interface {
	Inject(*substate)
}

// InjectorFunc defines a convenience type making it easy to implement
// Injectors.
type InjectorFunc func(*substate)

// Inject implements the Injector interface.
func (fn InjectorFunc) Inject(s *substate) {
	fn(s)
}

func WithFooClient(fooClient FooClient) InjectorFunc {
	return func(s *substate) {
		s.fooClient = fooClient
	}
}

func WithBarClient(barClient BarClient) InjectorFunc {
	return func(s *substate) {
		s.barClient = barClient
	}
}

type substate struct {
	fooClient FooClient
	barClient BarClient
}

func (s *substate) FooClient() FooClient {
	return s.fooClient
}
func (s *substate) BarClient() BarClient {
	return s.barClient
}
